### PR TITLE
Push/pop group operations (take 2)

### PR DESCRIPTION
**Pull request by Rubo3**
At the moment it's (nearly) the same as #87.

### CODE REVIEW

**Comment by Rubo3** [comment]
Here https://github.com/jpbruyere/vkvg/pull/87#issuecomment-1048585457 did you actually mean that it would suffice to just add a `VkvgSurface surf` member to the `vkvg_context_save_t` struct? (Clearly allocating and freeing it when necessary)

**Review by jpbruyere — COMMENTED** [review]
Do a rebase master, and forced push when you want the latest commit from master, I don't know if this merge would fit with the main git history, maybe yes...

**Comment by Rubo3** [comment]
Why do you think it?
Also, I updated push and pop to make use of save and restore, am I thinking it right? I still have to make save and restore update `ctx->pSurf`, for now assume they work like that.

File: src/vkvg_context.c [hunk 1]
```diff
@@ -1547,3 +1584,61 @@ void vkvg_ellipse (VkvgContext ctx, float radiusX, float radiusY, float x, float
...
+    if (ctx->status)
+        return NULL;
+    return ctx->pSurf;
+}
+
+void vkvg_push_group (VkvgContext ctx) {
+    if (ctx->status)
+        return;
+
+    vkvg_flush(ctx);
```
Comment (jpbruyere) [inline src/vkvg_context.c:1598]: context is flushed on save, so no need to flush here

File: src/vkvg_context.c [hunk 2]
```diff
@@ -520,6 +522,12 @@ void vkvg_move_to (VkvgContext ctx, float x, float y)
 	_finish_path(ctx);
 	_add_point (ctx, x, y);
 }
+
+bool vkvg_has_current_point (VkvgContext ctx) {
+    if (ctx->status)
+        return false;
+    return true;
```
Comment (jpbruyere) [inline src/vkvg_context.c]: did you notice there's a missing '}' here????

**Comment by Rubo3** [comment]
I removed roughly all my indentation changes.

**Comment by Rubo3** [comment]
I'm not sure I completely understood https://github.com/jpbruyere/vkvg/pull/87#issuecomment-1047872273. Can you please expand on it?

**Comment by jpbruyere** [comment]
> I'm not sure I completely understood [#87 (comment)](https://github.com/jpbruyere/vkvg/pull/87#issuecomment-1047872273). Can you please expand on it?

Internally, there is no new pattern created when you call `vkvg_set_source_rgb` and `vkvg_set_source_rgba`, the current color is stored in a separate field of the context (ctx->curColor). This is an optimization for fast solid color changes. But from the user point of view, when you set a solid color, a new solid color pattern is set for the context. This pattern has to be created only if the user make a call to `vkvg_get_source`. Once the user made a call to get the current pattern while it's a solid color, the pattern is created and must be used internally as any other pattern (and correctly referenced and destroyed).

**Comment by Rubo3** [comment]
> This pattern has to be created only if the user make a call to vkvg_get_source. Once the user made a call to get the current pattern while it's a solid color, the pattern is created

Got that, but I think I'm missing something: `vkvg_get_source` currently expects the pattern pointer to be non-NULL, otherwise `vkvg_pattern_reference` segfaults, so are you saying that the pattern creation needs to be added to `vkvg_get_source`?

**Pull request closed without merging**